File: crates/swc_ecma_transforms/scripts/del.py
#!/usr/bin/python3


import shutil
from os import listdir
from os.path import isfile, join


def search(dir):
    print('Searching ', dir)
    entries = listdir(dir)
    if len(entries) == 1 and 'fixtures' in dir:
        shutil.rmtree(dir, ignore_errors=True)
        return True

    if len(entries) == 2 and 'fixtures' in dir and 'options.json' in entries and 'input.js' in entries:
        shutil.rmtree(dir, ignore_errors=True)
        return True


    shouldDelete = True;
    for f in listdir(dir):
        p = join(dir, f);
        if isfile(p):
            if f == 'options.json':
                return False
        else:
            if not search(p):
                shouldDelete = False
            if isfile(join(dir, 'options.json')) and not isfile(join(p, 'options.json')):
                shutil.copyfile(join(dir, 'options.json'), join(p, 'options.json'))
    if shouldDelete and 'fixtures' in dir:
        shutil.rmtree(dir, ignore_errors=True)
    return True


search('./')

File: crates/swc_ecma_transforms/scripts/del.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

/**
 * Propagate fixture options downward before pruning empty fixture branches.
 */
function search(dirPath) {
  console.log(`Searching ${dirPath}`);

  const entries = fs.readdirSync(dirPath, { withFileTypes: true });
  const entryNames = entries.map((entry) => entry.name);
  const isFixtureDir = dirPath.includes('fixtures');

  if (isFixtureDir && entryNames.length === 1) {
    fs.rmSync(dirPath, { force: true, recursive: true });
    return true;
  }

  if (
    isFixtureDir &&
    entryNames.length === 2 &&
    entryNames.includes('options.json') &&
    entryNames.includes('input.js')
  ) {
    fs.rmSync(dirPath, { force: true, recursive: true });
    return true;
  }

  const localOptionsPath = path.join(dirPath, 'options.json');
  const hasLocalOptions = fs.existsSync(localOptionsPath);
  let shouldDelete = true;

  for (const entry of entries) {
    if (!entry.isDirectory()) {
      continue;
    }

    const childDir = path.join(dirPath, entry.name);
    const childContainsOptions = !search(childDir);
    if (childContainsOptions) {
      shouldDelete = false;
    }

    const childOptionsPath = path.join(childDir, 'options.json');
    if (hasLocalOptions && fs.existsSync(childDir) && !fs.existsSync(childOptionsPath)) {
      fs.copyFileSync(localOptionsPath, childOptionsPath);
    }
  }

  if (hasLocalOptions) {
    shouldDelete = false;
  }

  if (shouldDelete && isFixtureDir) {
    fs.rmSync(dirPath, { force: true, recursive: true });
  }

  return !hasLocalOptions;
}

search('.');
